test/com/amazon/ion/impl/IonReaderContinuableTopLevelBinaryTest.java

```diff
@@ -1813,6 +1856,995 @@ public void symbolTableSnapshotImplementsBasicMethods() throws Exception {
+     */
+    private IonReader boundedReaderFor(InputStream stream, int initialBufferSize, int maximumBufferSize, UnifiedTestHandler handler) {
+        byteCounter.set(0);
+        setBufferBounds(initialBufferSize, maximumBufferSize, handler);
+        return readerFor(stream);
+    }
+
+    @Test
+    public void singleValueExceedsInitialBufferSize() throws Exception {
+        IonReader reader = boundedReaderFor(
+            toBinary("\"abcdefghijklmnopqrstuvwxyz\""),
+            8,
+            Integer.MAX_VALUE,
+            byteCountingHandler
+        );
+        assertEquals(IonType.STRING, reader.next());
+        assertEquals("abcdefghijklmnopqrstuvwxyz", reader.stringValue());
+        assertNull(reader.next());
+        reader.close();
+        assertBytesConsumed();
+    }
+
+    @Test
+    public void maximumBufferSizeTooSmallFails() {
+        IonBufferConfiguration.Builder builder = IonBufferConfiguration.Builder.standard();
+        builder
+            .withMaximumBufferSize(builder.getMinimumMaximumBufferSize() - 1)
+            .withInitialBufferSize(builder.getMinimumMaximumBufferSize() - 1)
+            .onOversizedValue(builder.getNoOpOversizedValueHandler())
+            .onOversizedSymbolTable(builder.getNoOpOversizedSymbolTableHandler())
+            .onData(builder.getNoOpDataHandler());
+        thrown.expect(IllegalArgumentException.class);
+        builder.build();
+    }
+
+    @Test
+    public void maximumBufferSizeWithoutHandlerFails() {
+        IonBufferConfiguration.Builder builder = IonBufferConfiguration.Builder.standard();
+        builder
+            .withMaximumBufferSize(9)
+            .withInitialBufferSize(9);
+        thrown.expect(IllegalArgumentException.class);
+        builder.build();
+    }
+
+    @Test
+    public void oversizeValueDetectedDuringScalarFill() throws Exception {
+        byte[] bytes = toBinary(
+            "\"abcdefghijklmnopqrstuvwxyz\" " + // Requires 32 bytes (4 IVM, 1 TID, 1 length, 26 chars)
+            "\"abc\" " +
+            "\"abcdefghijklmnopqrstuvwxyz\" " +
+            "\"def\""
+        );
+
+        IonReader reader = boundedReaderFor(new ByteArrayInputStream(bytes), 8, 16, byteAndOversizedValueCountingHandler);
+
+        assertEquals(IonType.STRING, reader.next());
+        assertEquals("abc", reader.stringValue());
+        assertEquals(1, oversizedCounter.get());
+        assertEquals(IonType.STRING, reader.next());
+        assertEquals("def", reader.stringValue());
+        assertNull(reader.next());
+        reader.close();
+        assertEquals(2, oversizedCounter.get());
+        totalBytesInStream = bytes.length;
+        assertBytesConsumed();
+    }
+
+    @Test
+    public void oversizeValueDetectedDuringScalarFillIncremental() throws Exception {
+        byte[] bytes = toBinary(
+            "\"abcdefghijklmnopqrstuvwxyz\" " + // Requires 32 bytes (4 IVM, 1 TID, 1 length, 26 chars)
```

Comment:
do the 4 IVM bytes count towards the limit?

Reply:
Yes, for simplicity we buffer the entire IVM so that we don't have to have slow/unchecked variants for parsing the IVM.